File: packages/sdk-python/rollgate/metrics.py
"""
SDK Metrics Collection.
Tracks request performance, cache efficiency, error rates, and flag evaluations.
"""

import time
from dataclasses import dataclass, field
from typing import Dict, List, Optional, Callable, Any
from enum import Enum
from collections import defaultdict


class CircuitStateValue(Enum):
    """Circuit breaker state values."""
    CLOSED = "closed"
    OPEN = "open"
    HALF_OPEN = "half-open"


@dataclass
class WindowedStats:
    """Statistics for a time window."""
    requests: int = 0
    errors: int = 0
    avg_latency_ms: float = 0
    error_rate: float = 0


@dataclass
class FlagStats:
    """Statistics for a single flag."""
    evaluations: int = 0
    true_count: int = 0
    false_count: int = 0
    true_rate: float = 0
    avg_evaluation_time_ms: float = 0


@dataclass
class FlagEvaluationMetrics:
    """Flag evaluation metrics."""
    total_evaluations: int = 0
    evaluations_per_flag: Dict[str, FlagStats] = field(default_factory=dict)
    avg_evaluation_time_ms: float = 0


@dataclass
class TimeWindowMetrics:
    """Time-windowed metrics."""
    one_minute: WindowedStats = field(default_factory=WindowedStats)
    five_minutes: WindowedStats = field(default_factory=WindowedStats)
    fifteen_minutes: WindowedStats = field(default_factory=WindowedStats)
    one_hour: WindowedStats = field(default_factory=WindowedStats)


@dataclass
class MetricsSnapshot:
    """Complete snapshot of all metrics."""
    # Request metrics
    total_requests: int = 0
    successful_requests: int = 0
    failed_requests: int = 0
    success_rate: float = 0
    error_rate: float = 0

    # Latency metrics (in milliseconds)
    avg_latency_ms: float = 0
    min_latency_ms: float = 0
    max_latency_ms: float = 0
    p50_latency_ms: float = 0
    p95_latency_ms: float = 0
    p99_latency_ms: float = 0

    # Cache metrics
    cache_hits: int = 0
    cache_misses: int = 0
    cache_hit_rate: float = 0
    not_modified_responses: int = 0

    # Error metrics
    errors_by_category: Dict[str, int] = field(default_factory=dict)

    # Circuit breaker metrics
    circuit_opens: int = 0
    circuit_closes: int = 0
    circuit_state: str = "closed"

    # Flag evaluation metrics
    flag_evaluations: FlagEvaluationMetrics = field(default_factory=FlagEvaluationMetrics)

    # Time-windowed metrics
    windows: TimeWindowMetrics = field(default_factory=TimeWindowMetrics)

    # Timing
    uptime_ms: int = 0
    last_request_at: Optional[int] = None


@dataclass
class RequestMetrics:
    """Metrics for a single request."""
    endpoint: str
    status_code: int
    latency_ms: float
    cache_hit: bool = False
    not_modified: bool = False
    error: Optional[str] = None
    error_category: Optional[str] = None


@dataclass
class TimestampedRequest:
    """Request with timestamp for time windows."""
    timestamp: float
    latency_ms: float
    success: bool


@dataclass
class FlagEvaluation:
    """Record of a flag evaluation."""
    flag_key: str
    result: bool
    evaluation_time_ms: float
    timestamp: float


# Time windows in seconds
TIME_WINDOWS = {
    "1m": 60,
    "5m": 5 * 60,
    "15m": 15 * 60,
    "1h": 60 * 60,
}


class SDKMetrics:
    """
    Collects and reports SDK metrics.

    Example:
        ```python
        metrics = SDKMetrics()

        # Record a request
        metrics.record_request(RequestMetrics(
            endpoint="/api/v1/flags",
            status_code=200,
            latency_ms=45.2,
            cache_hit=False,
        ))

        # Record a flag evaluation
        metrics.record_evaluation("my-feature", True, 0.5)

        # Get snapshot
        snap = metrics.snapshot()
        print(f"Success rate: {snap.success_rate}%")

        # Export to Prometheus format
        print(metrics.to_prometheus())
        ```
    """

    def __init__(self):
        """Initialize metrics collector."""
        self._total_requests = 0
        self._successful_requests = 0
        self._failed_requests = 0
        self._cache_hits = 0
        self._cache_misses = 0
        self._not_modified_responses = 0
        self._circuit_opens = 0
        self._circuit_closes = 0
        self._circuit_state = CircuitStateValue.CLOSED

        self._latencies: List[float] = []
        self._max_latency_history = 1000

        self._errors_by_category: Dict[str, int] = defaultdict(int)

        # Time-windowed request tracking
        self._timestamped_requests: List[TimestampedRequest] = []
        self._max_timestamped_requests = 10000

        # Flag evaluation tracking
        self._flag_stats: Dict[str, Dict[str, Any]] = {}
        self._timestamped_evaluations: List[FlagEvaluation] = []
        self._max_timestamped_evaluations = 10000
        self._total_evaluations = 0
        self._total_evaluation_time_ms = 0.0

        self._start_time = time.time()
        self._last_request_at: Optional[float] = None

        # Event listeners
        self._listeners: Dict[str, List[Callable[[MetricsSnapshot], None]]] = defaultdict(list)

    def record_request(self, metrics: RequestMetrics) -> None:
        """
        Record a completed request.

        Args:
            metrics: Request metrics to record
        """
        now = time.time()
        self._total_requests += 1
        self._last_request_at = now

        success = 200 <= metrics.status_code < 400

        if success:
            self._successful_requests += 1
        else:
            self._failed_requests += 1

        # Track cache metrics
        if metrics.not_modified:
            self._not_modified_responses += 1
            self._cache_hits += 1
        elif metrics.cache_hit:
            self._cache_hits += 1
        else:
            self._cache_misses += 1

        # Track latency
        self._latencies.append(metrics.latency_ms)
        if len(self._latencies) > self._max_latency_history:
            self._latencies.pop(0)

        # Track timestamped request for time windows
        self._timestamped_requests.append(TimestampedRequest(
            timestamp=now,
            latency_ms=metrics.latency_ms,
            success=success,
        ))
        if len(self._timestamped_requests) > self._max_timestamped_requests:
            self._timestamped_requests.pop(0)

        # Track errors by category
        if metrics.error_category:
            self._errors_by_category[metrics.error_category] += 1

        # Emit update event
        self._emit("request", self.snapshot())

    def record_evaluation(
        self,
        flag_key: str,
        result: bool,
        evaluation_time_ms: float = 0,
    ) -> None:
        """
        Record a flag evaluation.

        Args:
            flag_key: The flag that was evaluated
            result: Evaluation result
            evaluation_time_ms: Time taken to evaluate
        """
        now = time.time()
        self._total_evaluations += 1
        self._total_evaluation_time_ms += evaluation_time_ms

        # Update per-flag stats
        if flag_key not in self._flag_stats:
            self._flag_stats[flag_key] = {
                "count": 0,
                "true_count": 0,
                "total_time_ms": 0,
            }

        stats = self._flag_stats[flag_key]
        stats["count"] += 1
        if result:
            stats["true_count"] += 1
        stats["total_time_ms"] += evaluation_time_ms

        # Track timestamped evaluation for time windows
        self._timestamped_evaluations.append(FlagEvaluation(
            flag_key=flag_key,
            result=result,
            evaluation_time_ms=evaluation_time_ms,
            timestamp=now,
        ))
        if len(self._timestamped_evaluations) > self._max_timestamped_evaluations:
            self._timestamped_evaluations.pop(0)

        # Emit update event
        self._emit("evaluation", self.snapshot())

    def record_circuit_state_change(self, new_state: CircuitStateValue) -> None:
        """
        Record a circuit breaker state change.

        Args:
            new_state: New circuit state
        """
        old_state = self._circuit_state
        self._circuit_state = new_state

        if new_state == CircuitStateValue.OPEN and old_state != CircuitStateValue.OPEN:
            self._circuit_opens += 1
        elif new_state == CircuitStateValue.CLOSED and old_state != CircuitStateValue.CLOSED:
            self._circuit_closes += 1

        self._emit("circuit-change", self.snapshot())

    def get_circuit_state(self) -> CircuitStateValue:
        """Get current circuit breaker state."""
        return self._circuit_state

    def on(
        self,
        event: str,
        callback: Callable[[MetricsSnapshot], None],
    ) -> None:
        """
        Subscribe to metrics events.

        Args:
            event: Event name ('request', 'evaluation', 'circuit-change')
            callback: Callback function
        """
        self._listeners[event].append(callback)

    def off(
        self,
        event: str,
        callback: Callable[[MetricsSnapshot], None],
    ) -> None:
        """
        Unsubscribe from metrics events.

        Args:
            event: Event name
            callback: Callback function to remove
        """
        if callback in self._listeners[event]:
            self._listeners[event].remove(callback)

    def _emit(self, event: str, data: MetricsSnapshot) -> None:
        """Emit an event to all listeners."""
        for callback in self._listeners.get(event, []):
            try:
                callback(data)
            except Exception:
                pass  # Ignore callback errors

    def snapshot(self) -> MetricsSnapshot:
        """
        Get a snapshot of all metrics.

        Returns:
            Complete metrics snapshot
        """
        sorted_latencies = sorted(self._latencies)
        total_cache_requests = self._cache_hits + self._cache_misses

        return MetricsSnapshot(
            total_requests=self._total_requests,
            successful_requests=self._successful_requests,
            failed_requests=self._failed_requests,
            success_rate=(
                (self._successful_requests / self._total_requests) * 100
                if self._total_requests > 0 else 0
            ),
            error_rate=(
                (self._failed_requests / self._total_requests) * 100
                if self._total_requests > 0 else 0
            ),

            avg_latency_ms=self._calculate_average(sorted_latencies),
            min_latency_ms=sorted_latencies[0] if sorted_latencies else 0,
            max_latency_ms=sorted_latencies[-1] if sorted_latencies else 0,
            p50_latency_ms=self._calculate_percentile(sorted_latencies, 50),
            p95_latency_ms=self._calculate_percentile(sorted_latencies, 95),
            p99_latency_ms=self._calculate_percentile(sorted_latencies, 99),

            cache_hits=self._cache_hits,
            cache_misses=self._cache_misses,
            cache_hit_rate=(
                (self._cache_hits / total_cache_requests) * 100
                if total_cache_requests > 0 else 0
            ),
            not_modified_responses=self._not_modified_responses,

            errors_by_category=dict(self._errors_by_category),

            circuit_opens=self._circuit_opens,
            circuit_closes=self._circuit_closes,
            circuit_state=self._circuit_state.value,

            flag_evaluations=self._get_flag_evaluation_metrics(),
            windows=self._get_time_window_metrics(),

            uptime_ms=int((time.time() - self._start_time) * 1000),
            last_request_at=(
                int(self._last_request_at * 1000)
                if self._last_request_at else None
            ),
        )

    def _get_flag_evaluation_metrics(self) -> FlagEvaluationMetrics:
        """Get flag evaluation metrics."""
        evaluations_per_flag: Dict[str, FlagStats] = {}

        for flag_key, stats in self._flag_stats.items():
            count = stats["count"]
            true_count = stats["true_count"]
            evaluations_per_flag[flag_key] = FlagStats(
                evaluations=count,
                true_count=true_count,
                false_count=count - true_count,
                true_rate=(true_count / count) * 100 if count > 0 else 0,
                avg_evaluation_time_ms=(
                    stats["total_time_ms"] / count if count > 0 else 0
                ),
            )

        return FlagEvaluationMetrics(
            total_evaluations=self._total_evaluations,
            evaluations_per_flag=evaluations_per_flag,
            avg_evaluation_time_ms=(
                self._total_evaluation_time_ms / self._total_evaluations
                if self._total_evaluations > 0 else 0
            ),
        )

    def _get_time_window_metrics(self) -> TimeWindowMetrics:
        """Get time-windowed metrics."""
        now = time.time()
        return TimeWindowMetrics(
            one_minute=self._calculate_window_stats(now, TIME_WINDOWS["1m"]),
            five_minutes=self._calculate_window_stats(now, TIME_WINDOWS["5m"]),
            fifteen_minutes=self._calculate_window_stats(now, TIME_WINDOWS["15m"]),
            one_hour=self._calculate_window_stats(now, TIME_WINDOWS["1h"]),
        )

    def _calculate_window_stats(self, now: float, window_seconds: float) -> WindowedStats:
        """Calculate stats for a time window."""
        cutoff = now - window_seconds
        window_requests = [r for r in self._timestamped_requests if r.timestamp >= cutoff]

        requests = len(window_requests)
        errors = sum(1 for r in window_requests if not r.success)
        total_latency = sum(r.latency_ms for r in window_requests)

        return WindowedStats(
            requests=requests,
            errors=errors,
            avg_latency_ms=total_latency / requests if requests > 0 else 0,
            error_rate=(errors / requests) * 100 if requests > 0 else 0,
        )

    def to_prometheus(self, prefix: str = "rollgate_sdk") -> str:
        """
        Export metrics in Prometheus format.

        Args:
            prefix: Metric name prefix

        Returns:
            Metrics in Prometheus text format
        """
        snap = self.snapshot()
        lines: List[str] = []

        def metric(name: str, value: float, help_text: str = "", metric_type: str = ""):
            full_name = f"{prefix}_{name}"
            if help_text:
                lines.append(f"# HELP {full_name} {help_text}")
            if metric_type:
                lines.append(f"# TYPE {full_name} {metric_type}")
            lines.append(f"{full_name} {value}")

        # Request metrics
        metric("requests_total", snap.total_requests, "Total number of requests", "counter")
        metric("requests_success_total", snap.successful_requests, "Total successful requests", "counter")
        metric("requests_failed_total", snap.failed_requests, "Total failed requests", "counter")

        # Latency metrics
        metric("latency_avg_ms", snap.avg_latency_ms, "Average request latency in milliseconds", "gauge")
        metric("latency_p50_ms", snap.p50_latency_ms, "50th percentile latency", "gauge")
        metric("latency_p95_ms", snap.p95_latency_ms, "95th percentile latency", "gauge")
        metric("latency_p99_ms", snap.p99_latency_ms, "99th percentile latency", "gauge")

        # Cache metrics
        metric("cache_hits_total", snap.cache_hits, "Total cache hits", "counter")
        metric("cache_misses_total", snap.cache_misses, "Total cache misses", "counter")
        metric("cache_hit_rate", snap.cache_hit_rate, "Cache hit rate percentage", "gauge")

        # Circuit breaker metrics
        metric("circuit_opens_total", snap.circuit_opens, "Total circuit breaker opens", "counter")
        circuit_value = 0 if snap.circuit_state == "closed" else (1 if snap.circuit_state == "open" else 0.5)
        metric("circuit_state", circuit_value, "Circuit breaker state (0=closed, 0.5=half-open, 1=open)", "gauge")

        # Flag evaluation metrics
        metric("evaluations_total", snap.flag_evaluations.total_evaluations, "Total flag evaluations", "counter")
        metric(
            "evaluation_avg_time_ms",
            snap.flag_evaluations.avg_evaluation_time_ms,
            "Average evaluation time in milliseconds",
            "gauge",
        )

        # Uptime
        metric("uptime_seconds", snap.uptime_ms / 1000, "SDK uptime in seconds", "gauge")

        return "\n".join(lines)

    def reset(self) -> None:
        """Reset all metrics."""
        self._total_requests = 0
        self._successful_requests = 0
        self._failed_requests = 0
        self._cache_hits = 0
        self._cache_misses = 0
        self._not_modified_responses = 0
        self._circuit_opens = 0
        self._circuit_closes = 0
        self._circuit_state = CircuitStateValue.CLOSED
        self._latencies = []
        self._errors_by_category = defaultdict(int)
        self._timestamped_requests = []
        self._flag_stats = {}
        self._timestamped_evaluations = []
        self._total_evaluations = 0
        self._total_evaluation_time_ms = 0.0
        self._start_time = time.time()
        self._last_request_at = None

    @staticmethod
    def _calculate_average(sorted_values: List[float]) -> float:
        """Calculate average of sorted values."""
        if not sorted_values:
            return 0
        return sum(sorted_values) / len(sorted_values)

    @staticmethod
    def _calculate_percentile(sorted_values: List[float], percentile: float) -> float:
        """Calculate percentile of sorted values."""
        if not sorted_values:
            return 0
        index = int((percentile / 100) * len(sorted_values)) - 1
        return sorted_values[max(0, index)]


# Global metrics instance
_global_metrics: Optional[SDKMetrics] = None


def get_metrics() -> SDKMetrics:
    """Get or create the global metrics instance."""
    global _global_metrics
    if _global_metrics is None:
        _global_metrics = SDKMetrics()
    return _global_metrics


def create_metrics() -> SDKMetrics:
    """Create a new metrics instance (useful for testing)."""
    return SDKMetrics()

File: packages/sdk-core/src/metrics.ts
/**
 * SDK Metrics Collection
 * Tracks request performance, cache efficiency, error rates, and flag evaluations
 */

// Time window durations in milliseconds
const TIME_WINDOWS = {
  "1m": 60 * 1000,
  "5m": 5 * 60 * 1000,
  "15m": 15 * 60 * 1000,
  "1h": 60 * 60 * 1000,
} as const;

export interface MetricsSnapshot {
  // Request metrics
  totalRequests: number;
  successfulRequests: number;
  failedRequests: number;
  successRate: number;
  errorRate: number;

  // Latency metrics (in milliseconds)
  avgLatencyMs: number;
  minLatencyMs: number;
  maxLatencyMs: number;
  p50LatencyMs: number;
  p95LatencyMs: number;
  p99LatencyMs: number;

  // Cache metrics
  cacheHits: number;
  cacheMisses: number;
  cacheHitRate: number;
  notModifiedResponses: number;

  // Error metrics
  errorsByCategory: Record<string, number>;

  // Circuit breaker metrics
  circuitOpens: number;
  circuitCloses: number;
  circuitState: "closed" | "open" | "half-open";

  // Flag evaluation metrics
  flagEvaluations: FlagEvaluationMetrics;

  // Time-windowed metrics
  windows: TimeWindowMetrics;

  // Timing
  uptimeMs: number;
  lastRequestAt: number | null;
}

export interface FlagEvaluationMetrics {
  totalEvaluations: number;
  evaluationsPerFlag: Record<string, FlagStats>;
  avgEvaluationTimeMs: number;
}

export interface FlagStats {
  evaluations: number;
  trueCount: number;
  falseCount: number;
  trueRate: number;
  avgEvaluationTimeMs: number;
}

export interface TimeWindowMetrics {
  "1m": WindowedStats;
  "5m": WindowedStats;
  "15m": WindowedStats;
  "1h": WindowedStats;
}

export interface WindowedStats {
  requests: number;
  errors: number;
  avgLatencyMs: number;
  errorRate: number;
}

export interface RequestMetrics {
  endpoint: string;
  statusCode: number;
  latencyMs: number;
  cacheHit: boolean;
  notModified: boolean;
  error?: string;
  errorCategory?: string;
}

export interface FlagEvaluationRecord {
  flagKey: string;
  result: boolean;
  evaluationTimeMs: number;
  timestamp: number;
}

interface TimestampedRequest {
  timestamp: number;
  latencyMs: number;
  success: boolean;
}

interface TimestampedEvaluation {
  flagKey: string;
  result: boolean;
  evaluationTimeMs: number;
  timestamp: number;
}

export class SDKMetrics {
  private totalRequests = 0;
  private successfulRequests = 0;
  private failedRequests = 0;
  private cacheHits = 0;
  private cacheMisses = 0;
  private notModifiedResponses = 0;
  private circuitOpens = 0;
  private circuitCloses = 0;
  private circuitState: "closed" | "open" | "half-open" = "closed";

  private latencies: number[] = [];
  private maxLatencyHistory = 1000;

  private errorsByCategory: Map<string, number> = new Map();

  // Time-windowed request tracking
  private timestampedRequests: TimestampedRequest[] = [];
  private maxTimestampedRequests = 10000;

  // Flag evaluation tracking
  private flagEvaluations: Map<
    string,
    { count: number; trueCount: number; totalTimeMs: number }
  > = new Map();
  private timestampedEvaluations: TimestampedEvaluation[] = [];
  private maxTimestampedEvaluations = 10000;
  private totalEvaluations = 0;
  private totalEvaluationTimeMs = 0;

  private startTime = Date.now();
  private lastRequestAt: number | null = null;

  // Event listeners for metrics updates
  private listeners: Map<string, Set<(metrics: MetricsSnapshot) => void>> =
    new Map();

  /**
   * Record a completed request
   */
  recordRequest(metrics: RequestMetrics): void {
    const now = Date.now();
    this.totalRequests++;
    this.lastRequestAt = now;

    const success = metrics.statusCode >= 200 && metrics.statusCode < 400;

    if (success) {
      this.successfulRequests++;
    } else {
      this.failedRequests++;
    }

    // Track cache metrics
    if (metrics.notModified) {
      this.notModifiedResponses++;
      this.cacheHits++;
    } else if (metrics.cacheHit) {
      this.cacheHits++;
    } else {
      this.cacheMisses++;
    }

    // Track latency
    this.latencies.push(metrics.latencyMs);
    if (this.latencies.length > this.maxLatencyHistory) {
      this.latencies.shift();
    }

    // Track timestamped request for time windows
    this.timestampedRequests.push({
      timestamp: now,
      latencyMs: metrics.latencyMs,
      success,
    });
    if (this.timestampedRequests.length > this.maxTimestampedRequests) {
      this.timestampedRequests.shift();
    }

    // Track errors by category
    if (metrics.errorCategory) {
      const current = this.errorsByCategory.get(metrics.errorCategory) || 0;
      this.errorsByCategory.set(metrics.errorCategory, current + 1);
    }

    // Emit update event
    this.emit("request", this.snapshot());
  }

  /**
   * Record a flag evaluation
   */
  recordEvaluation(
    flagKey: string,
    result: boolean,
    evaluationTimeMs: number = 0,
  ): void {
    const now = Date.now();
    this.totalEvaluations++;
    this.totalEvaluationTimeMs += evaluationTimeMs;

    // Update per-flag stats
    const existing = this.flagEvaluations.get(flagKey) || {
      count: 0,
      trueCount: 0,
      totalTimeMs: 0,
    };
    existing.count++;
    if (result) existing.trueCount++;
    existing.totalTimeMs += evaluationTimeMs;
    this.flagEvaluations.set(flagKey, existing);

    // Track timestamped evaluation for time windows
    this.timestampedEvaluations.push({
      flagKey,
      result,
      evaluationTimeMs,
      timestamp: now,
    });
    if (this.timestampedEvaluations.length > this.maxTimestampedEvaluations) {
      this.timestampedEvaluations.shift();
    }

    // Emit update event
    this.emit("evaluation", this.snapshot());
  }

  /**
   * Record a circuit breaker state change
   */
  recordCircuitStateChange(newState: "closed" | "open" | "half-open"): void {
    const oldState = this.circuitState;
    this.circuitState = newState;

    if (newState === "open" && oldState !== "open") {
      this.circuitOpens++;
    } else if (newState === "closed" && oldState !== "closed") {
      this.circuitCloses++;
    }

    this.emit("circuit-change", this.snapshot());
  }

  /**
   * Get current circuit breaker state
   */
  getCircuitState(): "closed" | "open" | "half-open" {
    return this.circuitState;
  }

  /**
   * Subscribe to metrics events
   */
  on(
    event: "request" | "evaluation" | "circuit-change",
    callback: (metrics: MetricsSnapshot) => void,
  ): void {
    if (!this.listeners.has(event)) {
      this.listeners.set(event, new Set());
    }
    this.listeners.get(event)!.add(callback);
  }

  /**
   * Unsubscribe from metrics events
   */
  off(
    event: "request" | "evaluation" | "circuit-change",
    callback: (metrics: MetricsSnapshot) => void,
  ): void {
    this.listeners.get(event)?.delete(callback);
  }

  /**
   * Remove all event listeners (for cleanup)
   */
  removeAllListeners(): void {
    this.listeners.clear();
  }

  private emit(event: string, data: MetricsSnapshot): void {
    this.listeners.get(event)?.forEach((callback) => callback(data));
  }

  /**
   * Get a snapshot of all metrics
   */
  snapshot(): MetricsSnapshot {
    const sortedLatencies = [...this.latencies].sort((a, b) => a - b);
    const totalCacheRequests = this.cacheHits + this.cacheMisses;

    return {
      totalRequests: this.totalRequests,
      successfulRequests: this.successfulRequests,
      failedRequests: this.failedRequests,
      successRate:
        this.totalRequests > 0
          ? (this.successfulRequests / this.totalRequests) * 100
          : 0,
      errorRate:
        this.totalRequests > 0
          ? (this.failedRequests / this.totalRequests) * 100
          : 0,

      avgLatencyMs: this.calculateAverage(sortedLatencies),
      minLatencyMs: sortedLatencies.length > 0 ? sortedLatencies[0] : 0,
      maxLatencyMs:
        sortedLatencies.length > 0
          ? sortedLatencies[sortedLatencies.length - 1]
          : 0,
      p50LatencyMs: this.calculatePercentile(sortedLatencies, 50),
      p95LatencyMs: this.calculatePercentile(sortedLatencies, 95),
      p99LatencyMs: this.calculatePercentile(sortedLatencies, 99),

      cacheHits: this.cacheHits,
      cacheMisses: this.cacheMisses,
      cacheHitRate:
        totalCacheRequests > 0
          ? (this.cacheHits / totalCacheRequests) * 100
          : 0,
      notModifiedResponses: this.notModifiedResponses,

      errorsByCategory: Object.fromEntries(this.errorsByCategory),

      circuitOpens: this.circuitOpens,
      circuitCloses: this.circuitCloses,
      circuitState: this.circuitState,

      flagEvaluations: this.getFlagEvaluationMetrics(),
      windows: this.getTimeWindowMetrics(),

      uptimeMs: Date.now() - this.startTime,
      lastRequestAt: this.lastRequestAt,
    };
  }

  /**
   * Get flag evaluation metrics
   */
  private getFlagEvaluationMetrics(): FlagEvaluationMetrics {
    const evaluationsPerFlag: Record<string, FlagStats> = {};

    for (const [flagKey, stats] of this.flagEvaluations.entries()) {
      evaluationsPerFlag[flagKey] = {
        evaluations: stats.count,
        trueCount: stats.trueCount,
        falseCount: stats.count - stats.trueCount,
        trueRate: stats.count > 0 ? (stats.trueCount / stats.count) * 100 : 0,
        avgEvaluationTimeMs:
          stats.count > 0 ? stats.totalTimeMs / stats.count : 0,
      };
    }

    return {
      totalEvaluations: this.totalEvaluations,
      evaluationsPerFlag,
      avgEvaluationTimeMs:
        this.totalEvaluations > 0
          ? this.totalEvaluationTimeMs / this.totalEvaluations
          : 0,
    };
  }

  /**
   * Get time-windowed metrics
   */
  private getTimeWindowMetrics(): TimeWindowMetrics {
    const now = Date.now();

    const result: TimeWindowMetrics = {
      "1m": this.calculateWindowStats(now, TIME_WINDOWS["1m"]),
      "5m": this.calculateWindowStats(now, TIME_WINDOWS["5m"]),
      "15m": this.calculateWindowStats(now, TIME_WINDOWS["15m"]),
      "1h": this.calculateWindowStats(now, TIME_WINDOWS["1h"]),
    };

    return result;
  }

  private calculateWindowStats(now: number, windowMs: number): WindowedStats {
    const cutoff = now - windowMs;
    const windowRequests = this.timestampedRequests.filter(
      (r) => r.timestamp >= cutoff,
    );

    const requests = windowRequests.length;
    const errors = windowRequests.filter((r) => !r.success).length;
    const totalLatency = windowRequests.reduce(
      (sum, r) => sum + r.latencyMs,
      0,
    );

    return {
      requests,
      errors,
      avgLatencyMs: requests > 0 ? totalLatency / requests : 0,
      errorRate: requests > 0 ? (errors / requests) * 100 : 0,
    };
  }

  /**
   * Export metrics in Prometheus format
   */
  toPrometheus(prefix: string = "rollgate_sdk"): string {
    const snap = this.snapshot();
    const lines: string[] = [];

    // Helper to format metric
    const metric = (
      name: string,
      value: number,
      help?: string,
      type?: string,
    ) => {
      const fullName = `${prefix}_${name}`;
      if (help) lines.push(`# HELP ${fullName} ${help}`);
      if (type) lines.push(`# TYPE ${fullName} ${type}`);
      lines.push(`${fullName} ${value}`);
    };

    // Request metrics
    metric(
      "requests_total",
      snap.totalRequests,
      "Total number of requests",
      "counter",
    );
    metric(
      "requests_success_total",
      snap.successfulRequests,
      "Total successful requests",
      "counter",
    );
    metric(
      "requests_failed_total",
      snap.failedRequests,
      "Total failed requests",
      "counter",
    );

    // Latency metrics
    metric(
      "latency_avg_ms",
      snap.avgLatencyMs,
      "Average request latency in milliseconds",
      "gauge",
    );
    metric(
      "latency_p50_ms",
      snap.p50LatencyMs,
      "50th percentile latency",
      "gauge",
    );
    metric(
      "latency_p95_ms",
      snap.p95LatencyMs,
      "95th percentile latency",
      "gauge",
    );
    metric(
      "latency_p99_ms",
      snap.p99LatencyMs,
      "99th percentile latency",
      "gauge",
    );

    // Cache metrics
    metric("cache_hits_total", snap.cacheHits, "Total cache hits", "counter");
    metric(
      "cache_misses_total",
      snap.cacheMisses,
      "Total cache misses",
      "counter",
    );
    metric(
      "cache_hit_rate",
      snap.cacheHitRate,
      "Cache hit rate percentage",
      "gauge",
    );

    // Circuit breaker metrics
    metric(
      "circuit_opens_total",
      snap.circuitOpens,
      "Total circuit breaker opens",
      "counter",
    );
    metric(
      "circuit_state",
      snap.circuitState === "closed"
        ? 0
        : snap.circuitState === "open"
          ? 1
          : 0.5,
      "Circuit breaker state (0=closed, 0.5=half-open, 1=open)",
      "gauge",
    );

    // Flag evaluation metrics
    metric(
      "evaluations_total",
      snap.flagEvaluations.totalEvaluations,
      "Total flag evaluations",
      "counter",
    );
    metric(
      "evaluation_avg_time_ms",
      snap.flagEvaluations.avgEvaluationTimeMs,
      "Average evaluation time in milliseconds",
      "gauge",
    );

    // Uptime
    metric(
      "uptime_seconds",
      snap.uptimeMs / 1000,
      "SDK uptime in seconds",
      "gauge",
    );

    return lines.join("\n");
  }

  /**
   * Reset all metrics
   */
  reset(): void {
    this.totalRequests = 0;
    this.successfulRequests = 0;
    this.failedRequests = 0;
    this.cacheHits = 0;
    this.cacheMisses = 0;
    this.notModifiedResponses = 0;
    this.circuitOpens = 0;
    this.circuitCloses = 0;
    this.circuitState = "closed";
    this.latencies = [];
    this.errorsByCategory.clear();
    this.timestampedRequests = [];
    this.flagEvaluations.clear();
    this.timestampedEvaluations = [];
    this.totalEvaluations = 0;
    this.totalEvaluationTimeMs = 0;
    this.startTime = Date.now();
    this.lastRequestAt = null;
  }

  private calculateAverage(sortedValues: number[]): number {
    if (sortedValues.length === 0) return 0;
    const sum = sortedValues.reduce((a, b) => a + b, 0);
    return sum / sortedValues.length;
  }

  private calculatePercentile(
    sortedValues: number[],
    percentile: number,
  ): number {
    if (sortedValues.length === 0) return 0;
    const index = Math.ceil((percentile / 100) * sortedValues.length) - 1;
    return sortedValues[Math.max(0, index)];
  }
}

// Default global instance
let globalMetrics: SDKMetrics | null = null;

/**
 * Get or create the global metrics instance
 */
export function getMetrics(): SDKMetrics {
  if (!globalMetrics) {
    globalMetrics = new SDKMetrics();
  }
  return globalMetrics;
}

/**
 * Create a new metrics instance (useful for testing)
 */
export function createMetrics(): SDKMetrics {
  return new SDKMetrics();
}
